cmd/chisel/cmd_debug_check_release_archives.go

```diff
@@ -0,0 +1,214 @@
+}
+
+var archiveOpen = archive.Open
+
+func (cmd *cmdDebugCheckReleaseArchives) Execute(args []string) error {
+	release, err := obtainRelease(cmd.Release)
+	if err != nil {
+		return err
+	}
+
+	archives := make(map[string]archive.Archive)
+	for archiveName, archiveInfo := range release.Archives {
+		openArchive, err := archiveOpen(&archive.Options{
+			Label:      archiveName,
+			Version:    archiveInfo.Version,
+			Arch:       cmd.Arch,
+			Suites:     archiveInfo.Suites,
+			Components: archiveInfo.Components,
+			Pro:        archiveInfo.Pro,
+			CacheDir:   cache.DefaultDir("chisel"),
+			PubKeys:    archiveInfo.PubKeys,
+		})
+		if err != nil {
+			if err == archive.ErrCredentialsNotFound {
+				logf("Archive %q ignored: credentials not found\n", archiveName)
+				continue
+			}
+			return err
+		}
+		archives[archiveName] = openArchive
+	}
+
+	type ownership struct {
+		Mode yamlMode `yaml:"mode"`
+		Link string   `yaml:"link"`
+		// Pkgs is a correspondence from archive name to package names.
+		Pkgs map[string][]string `yaml:"packages,flow"`
+	}
+
+	var orderedPkgs []string
+	for packageName := range release.Packages {
+		orderedPkgs = append(orderedPkgs, packageName)
+	}
+	slices.Sort(orderedPkgs)
+
+	directories := map[string][]ownership{}
+	for archiveName, archive := range archives {
+		logf("Processing archive %s...", archiveName)
+		for _, pkgName := range orderedPkgs {
+			if !archive.Exists(pkgName) {
+				continue
+			}
+			pkgReader, _, err := archive.Fetch(pkgName)
+			if err != nil {
+				return err
+			}
+			dataReader, err := deb.DataReader(pkgReader)
+			if err != nil {
+				return err
+			}
+			tarReader := tar.NewReader(dataReader)
+			for {
+				tarHeader, err := tarReader.Next()
+				if err == io.EOF {
+					break
+				}
+				if err != nil {
+					return err
+				}
+
+				path, ok := sanitizeTarPath(tarHeader.Name)
+				if !ok {
+					continue
+				}
+				// Make paths uniform: while directories always end in '/',
+				// symlinks don't.
+				path = strings.TrimSuffix(path, "/")
+
+				dir := directories[path]
+				found := false
+				// We look for a previous group that has the same entry in
+				// terms of mode, link, etc. and we add the package to the
+				// group. If there is none, we create a new one.
+				for i, o := range dir {
+					if tarHeader.Linkname != "" || o.Link != "" {
+						if tarHeader.Linkname == o.Link {
+							o.Pkgs[archiveName] = append(o.Pkgs[archiveName], pkgName)
+							dir[i] = o
+							found = true
+							break
+						}
+					} else {
+						if tarHeader.Mode == int64(o.Mode) {
+							o.Pkgs[archiveName] = append(o.Pkgs[archiveName], pkgName)
+							dir[i] = o
+							found = true
+							break
+						}
+					}
+				}
+				if !found {
+					dir = append(dir, ownership{
+						Mode: yamlMode(tarHeader.Mode),
+						Link: tarHeader.Linkname,
+						Pkgs: map[string][]string{archiveName: []string{pkgName}},
+					})
+					directories[path] = dir
```

Comment:
So every single path in Ubuntu will be a key in a map, that has a value, a structure, that has a map, with a key that is an archive name, and as a value a list, where the typically single entry is a package name. So a map, with a key/value entry, pointing to a list, with a map, with a key/value, pointing to a list. Gut feeling isn't great.